ql/src/java/org/apache/hadoop/hive/ql/exec/CommonMergeJoinOperator.java

```diff
@@ -302,7 +322,47 @@ public void process(Object row, int tag) throws HiveException {
 
     assert !nextKeyGroup;
     candidateStorage[tag].addRow(value);
+  }
+
+  private void emitExtensionRow(int tag, List<Object> value) throws HiveException {
+    extensionStorage[tag].addRow(value);
+    for (byte i = 0; i < order.length; i++) {
+      if (i == tag) {
+        storage[i] = extensionStorage[i];
+      } else {
+        putDummyOrEmpty(i);
+      }
+    }
+    checkAndGenObject();
+    extensionStorage[tag].clearRows();
+  }
 
+  /**
+   * Decides if the actual row must be an extension row.
+   *
+   * Extension rows are those which are not part of the inner-join.
+   * May not correctly identify all extension rows - but will remove trivially filtered ones.
```

Comment:
`May not correctly...` It is not clear what this means. Possibly an example could help clarify this better.
What happens in the case where it does not identify an extension row? Wrong results?

Reply:
> May not identify all extension rows

it could not identify non-trivial ones like that no rows could satisfy the `l.id=r.id and l.id+2*r.id=1` condition...
guess what...running:
```
select * from t_xy l full outer join t_y r on (l.id=r.id and l.s='y' and l.id+2*r.id=1);
```
resulted in incorrect results as well...

this will leave me no choice but to remove that `skip` crap and replace it with something decent....

Reply:
Since you have examples that do not work, maybe it would be better to write "The current implementation, does not correctly identify all..." and put a reference to a follow up JIRA highlighting that there is an implementation limitation.